keps/sig-storage/20190129-csi-volume-resizing.md

```diff
@@ -0,0 +1,167 @@
+
+If `NodeExpandVolume` failed:
+- It will add a event to both PVC and Pod about failed resizing and resize operation will be retried. This will prevent pod from starting up.
+
+
+###### Online volume resizing on kubelet:
+
+More details about online resizing can be found in [Online resizing kep(https://github.com/kubernetes/enhancements/pull/737)] but essentially if
+`ExpandInUsePersistentVolumes` feature(currently Alpha) is enabled then kubelet will periodically poll all PVCs that are being used on the node and compare `pvc.spec.resources.requests.storage` and `pvc.Status.Capacity`(also `pv.Spec.Capacity`) and make similar determination about whether node expansion is required for the volume.
+
+In this mode `NodeExpandVolume` can be called while pod is running and volume is in-use. Using aformentioned check if kubelet determines that
+volume expansion is needed on the node and plugin implements `NodeExpandVolume` RPC call then, kubelet will call it(provided volume has already been node staged and published on the node) and:
+
+If `NodeExpandVolume` is successful:
+- It will update `pvc.Status.Capacity` with latest value and remove all resizing related conditions from PVC.
+
+If `NodeExpandVolume` failed:
+- It will add a event to both PVC and Pod about failed resizing and resize operation will be retried.
+
+### Risks and Mitigations
+
+Before this feature goes GA - we need to handle recovering https://github.com/kubernetes/kubernetes/issues/73036.
+
+## Test Plan
+
+* Unit tests for external resize controller.
+* Add e2e tests in Kubernetes that use csi-mock driver for volume resizing.
+  - (postive) Give a plugin that supports both control plane and node size resize, CSI volume should be resizable and able to complete successfully.
+  - (positive) Given a plugin that only requires control plane resize, CSI volume should be resizable and able to complete successfully.
+  - (positive) Given a plugin that only requires node side resize, CSI volume should be resizable and able to complete successfully.
+  - (negative) If control resize fails, PVC should have appropriate events.
+  - (neative) if node side resize fails, both pod and PVC should have appropriate events.
+
+## Graduation Criteria
+
+Once implemented CSI volumes should be resizable and in-line with current in-tree implementation of volume resizing.
+
+- *Alpha* : Initial support for CSI volume resizing. Released code will include an external CSI volume resize controller and changes to Kubelet. Implementation will have unit tests and csi-mock driver e2e tests.
+- *Beta*  : More robust support for CSI volume resizing. Add e2e tests that use real drivers(`gce-pd`, `ebs` at minimum). Add metrics for volume resize operations.
+- *GA* : CSI resizing in general will only leave GA after existing [Volume expansion(https://github.com/kubernetes/enhancements/issues/284)] feature leaves GA.
```

Comment:
should this also block on online resizing GA?

Reply:
I am not sure about that. I think this feature can go GA regardless of fate of online resizing.

Reply:
Can you clarify in the summary then that this kep is only for offline resizing support

Reply:
But this is just a individual driver implementation, whereas online and offline are resizing mechanisms. I consider this no different than say EBS in-tree resizing. Does EBS in-tree driver only targets offline resizing mechanism - it doesn't! So, IMO resizing mechanism going GA is different from resizing as a whole going GA and those are two different things.

Reply:
Say if by the time this feature is ready to go GA, but may be online resizing isn't (although that will not happen realistically) then we this feature go GA with default resizing mode. Online resizing will still be protected by its own feature gate and has to be explicitly enabled.